spotguide/spotguide.go

```diff
@@ -132,44 +126,112 @@ func (r LaunchRequest) RepoFullname() string {
-	reader, err := githubClient.Repositories.DownloadContents(ctx, owner, repo, file, &github.RepositoryContentGetOptions{
-		Ref: tag,
-	})
+type ReleaseBody struct {
+	Pipeline string `json:"pipeline" yaml:"pipeline"`
+}
+
+// SpotguideManager is responsible to scrape spotguides on GitHub and persist them to database
+type SpotguideManager struct {
+	pipelineVersion       *semver.Version
+	githubToken           string
+	spotguideOrganization *auth.Organization
+}
+
+func NewSpotguideManager(pipelineVersionString string) *SpotguideManager {
+	githubToken := viper.GetString("github.token")
+
+	spotguideOrganization, err := auth.GetOrganizationByName(SpotguideGithubOrganization)
 	if err != nil {
-		return nil, err
+		log.Errorf("shared spotguide organization (%s) is not found", SpotguideGithubOrganization)
 	}
 
-	defer reader.Close()
+	pipelineVersion, _ := semver.NewVersion(pipelineVersionString)
 
-	return ioutil.ReadAll(reader)
-}
-
-func internalScrapeSpotguides(orgID uint, userID uint) {
-	if err := ScrapeSpotguides(orgID, userID); err != nil {
-		log.Warnf("failed to scrape Spotguide repositories for org [%d]: %s", orgID, err)
+	s := &SpotguideManager{
+		pipelineVersion:       pipelineVersion,
+		spotguideOrganization: spotguideOrganization,
+		githubToken:           githubToken,
 	}
+
+	// subscribe to organization creations and sync spotguides into the newly created organizations
+	authEventEmitter.NotifyOrganizationRegistered(s.internalScrapeSpotguides)
+
+	// periodically sync shared spotguides
+	syncTicker := time.NewTicker(viper.GetDuration(config.SpotguideSyncInterval))
+	go func() {
+		if err := s.ScrapeSharedSpotguides(); err != nil {
+			log.Errorf("failed to sync shared spotguides: %v", err)
+		}
+
+		for range syncTicker.C {
+			if err := s.ScrapeSharedSpotguides(); err != nil {
+				log.Errorf("failed to sync shared spotguides: %v", err)
+			}
+		}
+	}()
+
+	return s
 }
 
-func isSpotguideReleaseAllowed(release *github.RepositoryRelease) bool {
+func (s *SpotguideManager) isSpotguideReleaseAllowed(release *github.RepositoryRelease) bool {
 	version, err := semver.NewVersion(release.GetTagName())
 	if err != nil {
-		log.Warn("Failed to parse spotguide release tag: ", err)
+		log.Warn("failed to parse spotguide release tag: ", err)
 		return false
 	}
-	return version.Prerelease() == "" || viper.GetBool(config.SpotguideAllowPrereleases)
+
+	supported := true
+	prerelease := version.Prerelease() != "" || *release.Prerelease
+
+	// try to parse release body as YAML
+	rawBody := release.GetBody()
+	body := ReleaseBody{}
+	err = yaml2.Unmarshal([]byte(rawBody), &body)
+	if s.pipelineVersion != nil && err == nil {
+		// check whether this release has support for this pipeline version
+		supportedConstraint, err := semver.NewConstraint(body.Pipeline)
+		if err == nil {
+			supported = supportedConstraint.Check(s.pipelineVersion)
+		}
+	}
+
+	return supported && (!prerelease || viper.GetBool(config.SpotguideAllowPrereleases))
```

Comment:
Try to avoid using global viper and rather inject these configuration values as well.